File: backend/app/routers/deaths.py
"""
Deaths router: death report submission and retrieval (farm-scoped).
"""

from datetime import date
from typing import Optional

from fastapi import (
    APIRouter,
    Depends,
    File,
    Form,
    HTTPException,
    Query,
    Request,
    UploadFile,
    status,
)
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.db_models import Animal, DeathRecord, User, utcnow
from app.models.schemas import DeathRecordResponse, Page
from app.services.audit_service import record_audit
from app.services.auth_service import get_current_user
from app.services.image_service import image_hash_exists, process_death_image
from app.services.storage import get_storage_backend
from app.utils import csv_response

router = APIRouter(prefix="/api/deaths", tags=["deaths"])


def _to_response(record: DeathRecord) -> DeathRecordResponse:
    """Serialize a record, resolving the storage reference to a public URL."""
    resp = DeathRecordResponse.model_validate(record)
    resp.image_path = get_storage_backend().public_url(record.image_path)
    return resp


@router.get("", response_model=Page[DeathRecordResponse])
def list_deaths(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
):
    """Return this farm's death reports (managers: all; employees: own)."""
    query = db.query(DeathRecord).filter(DeathRecord.farm_id == current_user.farm_id)
    if current_user.role != "manager":
        query = query.filter(DeathRecord.reported_by_user_id == current_user.id)
    total = query.count()
    records = (
        query.order_by(DeathRecord.created_at.desc())
        .offset((page - 1) * limit)
        .limit(limit)
        .all()
    )
    return Page(
        items=[_to_response(r) for r in records],
        total=total,
        page=page,
        limit=limit,
    )


@router.get("/check-hash")
def check_hash(
    hash: str = Query(..., min_length=64, max_length=64),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Report whether an image hash is already used in this farm.

    Lets the offline outbox warn a worker *before* they queue a duplicate.
    """
    return {"exists": image_hash_exists(db, current_user.farm_id, hash)}


@router.post(
    "", response_model=DeathRecordResponse, status_code=status.HTTP_201_CREATED
)
async def report_death(
    request: Request,
    animal_id: int = Form(...),
    cause_of_death: str = Form(...),
    date_of_death: date = Form(...),
    notes: Optional[str] = Form(None),
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Submit a death report for an animal in the caller's farm.

    - The animal must currently be alive.
    - A photo is required and must not have been used in a previous report
      within this farm (SHA-256 hash check).
    - On success the animal status is set to 'dead'.
    """
    animal = (
        db.query(Animal)
        .filter(Animal.id == animal_id, Animal.farm_id == current_user.farm_id)
        .first()
    )
    if not animal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        )

    if animal.status != "alive":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Animal '{animal.name}' is already recorded as dead",
        )

    existing_record = (
        db.query(DeathRecord).filter(DeathRecord.animal_id == animal_id).first()
    )
    if existing_record:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="A death record already exists for this animal",
        )

    # Validates file type/size and dedups within the farm; raises 409 on dup.
    image_ref, image_hash = await process_death_image(file, db, current_user.farm_id)

    record = DeathRecord(
        farm_id=current_user.farm_id,
        animal_id=animal_id,
        reported_by_user_id=current_user.id,
        cause_of_death=cause_of_death,
        date_of_death=date_of_death,
        image_path=image_ref,
        image_hash=image_hash,
        notes=notes,
    )
    db.add(record)

    animal.status = "dead"
    animal.updated_at = utcnow()

    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        # DB unique constraints (one record per animal, per-farm image hash)
        # are authoritative under concurrency. If a racing request beat us,
        # drop the object we just wrote and report 409.
        get_storage_backend().delete(image_ref)
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="A death report for this animal or image already exists",
        )
    db.refresh(record)
    record_audit(
        db,
        request,
        current_user,
        action="create",
        entity_type="death_record",
        entity_id=record.id,
        details={"animal_id": animal_id, "cause_of_death": cause_of_death},
    )
    return _to_response(record)


@router.get("/export.csv")
def export_deaths_csv(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Download this farm's death reports as CSV (employees: own only)."""
    query = db.query(DeathRecord).filter(DeathRecord.farm_id == current_user.farm_id)
    if current_user.role != "manager":
        query = query.filter(DeathRecord.reported_by_user_id == current_user.id)
    records = query.order_by(DeathRecord.created_at.desc()).all()
    header = [
        "id",
        "animal_id",
        "reported_by_user_id",
        "cause_of_death",
        "date_of_death",
        "notes",
        "created_at",
    ]
    rows = (
        [
            r.id,
            r.animal_id,
            r.reported_by_user_id,
            r.cause_of_death,
            r.date_of_death,
            r.notes or "",
            r.created_at,
        ]
        for r in records
    )
    return csv_response("deaths.csv", header, rows)


@router.get("/{record_id}", response_model=DeathRecordResponse)
def get_death(
    record_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return a single death report. Employees can only view their own."""
    record = (
        db.query(DeathRecord)
        .filter(
            DeathRecord.id == record_id,
            DeathRecord.farm_id == current_user.farm_id,
        )
        .first()
    )
    if not record:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Death record not found"
        )

    if current_user.role != "manager" and record.reported_by_user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You do not have permission to view this record",
        )
    return _to_response(record)

File: supabase/functions/api/routers/deaths.ts
// Port of backend/app/routers/deaths.py.
//
// Employee-self-only restriction (list/export/get): employees only ever see
// their OWN death reports; managers see every report in the farm. This is
// finer-grained than plain farm scoping — every read query below adds an
// `and reported_by_user_id = <employee id>` clause when the caller is not a
// manager, mirroring the SQLAlchemy `.filter(...)` chains in the Python
// router exactly.
import { Hono } from "hono";
import { getDb, isUniqueViolation } from "../../_shared/db.ts";
import { parsePagination } from "../../_shared/pagination.ts";
import {
  csvResponse,
  jsonError,
  pageEnvelope,
} from "../../_shared/response.ts";
import { recordAudit } from "../../_shared/auditService.ts";
import { requireAuth } from "../../_shared/auth.ts";
import {
  DeathRecordFormSchema,
  firstZodError,
} from "../../_shared/validation.ts";
import {
  DuplicateImageError,
  imageHashExists,
  ImageValidationError,
  processDeathImage,
} from "../../_shared/imageService.ts";
import {
  deleteObject,
  saveObject,
  signedUrl,
} from "../../_shared/storageService.ts";

interface DeathRecordRow {
  id: number;
  farm_id: number;
  animal_id: number;
  reported_by_user_id: number;
  cause_of_death: string;
  date_of_death: Date | string;
  image_path: string;
  image_hash: string;
  notes: string | null;
  created_at: Date | string;
}

interface DeathRecordResponse {
  id: number;
  farm_id: number;
  animal_id: number;
  reported_by_user_id: number;
  cause_of_death: string;
  date_of_death: string;
  image_path: string;
  notes: string | null;
  created_at: Date | string;
}

interface AnimalRow {
  id: number;
  name: string;
  status: "alive" | "dead";
}

// Postgres `date` columns come back as JS Date objects (UTC midnight); the
// Python response model serializes date_of_death as a bare "YYYY-MM-DD", so
// normalize the same way here rather than emitting a full ISO timestamp.
function toDateOnly(value: Date | string): string {
  if (typeof value === "string") return value.slice(0, 10);
  return value.toISOString().slice(0, 10);
}

export interface DeathsRouterDeps {
  /** Injection point for tests — defaults to the real Supabase Storage save. */
  saveFn?: typeof saveObject;
  /** Injection point for tests — defaults to the real Supabase Storage delete. */
  deleteFn?: typeof deleteObject;
  /** Injection point for tests — defaults to the real signed-URL resolver. */
  resolveUrl?: typeof signedUrl;
}

const HASH_LENGTH = 64;

export function createDeathsRouter(deps: DeathsRouterDeps = {}): Hono {
  const saveFn = deps.saveFn ?? saveObject;
  const deleteFn = deps.deleteFn ?? deleteObject;
  const resolveUrl = deps.resolveUrl ?? signedUrl;

  const router = new Hono();

  async function toResponse(
    record: DeathRecordRow,
  ): Promise<DeathRecordResponse> {
    return {
      id: record.id,
      farm_id: record.farm_id,
      animal_id: record.animal_id,
      reported_by_user_id: record.reported_by_user_id,
      cause_of_death: record.cause_of_death,
      date_of_death: toDateOnly(record.date_of_death),
      image_path: await resolveUrl(record.image_path),
      notes: record.notes,
      created_at: record.created_at,
    };
  }

  router.get("/check-hash", requireAuth, async (c) => {
    const hash = c.req.query("hash") ?? "";
    if (hash.length !== HASH_LENGTH) {
      return jsonError(
        c,
        422,
        `hash: String should have exactly ${HASH_LENGTH} characters`,
      );
    }
    const user = c.get("user");
    const exists = await imageHashExists(user.farm_id, hash);
    return c.json({ exists });
  });

  router.get("/export.csv", requireAuth, async (c) => {
    const user = c.get("user");
    const sql = getDb();
    const scopeFilter = user.role !== "manager"
      ? sql`and reported_by_user_id = ${user.id}`
      : sql``;
    const records = await sql<DeathRecordRow[]>`
      select * from death_records
      where farm_id = ${user.farm_id}
      ${scopeFilter}
      order by created_at desc
    `;
    const header = [
      "id",
      "animal_id",
      "reported_by_user_id",
      "cause_of_death",
      "date_of_death",
      "notes",
      "created_at",
    ];
    const rows = records.map((r) => [
      r.id,
      r.animal_id,
      r.reported_by_user_id,
      r.cause_of_death,
      toDateOnly(r.date_of_death),
      r.notes ?? "",
      r.created_at,
    ]);
    return csvResponse(c, "deaths.csv", header, rows);
  });

  router.get("", requireAuth, async (c) => {
    const user = c.get("user");
    const paged = parsePagination(c);
    if (!paged.ok) return jsonError(c, 422, paged.error);
    const { page, limit, offset } = paged.pagination;

    const sql = getDb();
    const scopeFilter = user.role !== "manager"
      ? sql`and reported_by_user_id = ${user.id}`
      : sql``;

    const [{ count }] = await sql<{ count: number }[]>`
      select count(*)::int as count from death_records
      where farm_id = ${user.farm_id}
      ${scopeFilter}
    `;
    const records = await sql<DeathRecordRow[]>`
      select * from death_records
      where farm_id = ${user.farm_id}
      ${scopeFilter}
      order by created_at desc
      offset ${offset} limit ${limit}
    `;
    const items = await Promise.all(records.map((r) => toResponse(r)));
    return c.json(pageEnvelope(items, count, page, limit));
  });

  router.post("", requireAuth, async (c) => {
    const user = c.get("user");

    const body = await c.req.parseBody();
    const file = body.file;
    if (!(file instanceof File)) {
      return jsonError(c, 422, "file: Field required");
    }

    const parsed = DeathRecordFormSchema.safeParse({
      animal_id: body.animal_id,
      cause_of_death: body.cause_of_death,
      date_of_death: body.date_of_death,
      notes: body.notes,
    });
    if (!parsed.success) return jsonError(c, 422, firstZodError(parsed));
    const {
      animal_id: animalId,
      cause_of_death: causeOfDeath,
      date_of_death: dateOfDeath,
      notes,
    } = parsed.data;

    const sql = getDb();

    const [animal] = await sql<AnimalRow[]>`
      select id, name, status from animals where id = ${animalId} and farm_id = ${user.farm_id}
    `;
    if (!animal) return jsonError(c, 404, "Animal not found");

    if (animal.status !== "alive") {
      return jsonError(
        c,
        400,
        `Animal '${animal.name}' is already recorded as dead`,
      );
    }

    const [existingRecord] = await sql<{ id: number }[]>`
      select id from death_records where animal_id = ${animalId}
    `;
    if (existingRecord) {
      return jsonError(c, 400, "A death record already exists for this animal");
    }

    const contents = new Uint8Array(await file.arrayBuffer());
    let imageRef: string;
    let imageHash: string;
    try {
      const result = await processDeathImage(
        file.type,
        contents,
        user.farm_id,
        saveFn,
      );
      imageRef = result.imageRef;
      imageHash = result.imageHash;
    } catch (err) {
      if (err instanceof DuplicateImageError) {
        return jsonError(c, 409, err.message);
      }
      if (err instanceof ImageValidationError) {
        return jsonError(c, 400, err.message);
      }
      throw err;
    }

    let record: DeathRecordRow;
    try {
      [record] = await sql.begin(async (tx) => {
        const [rec] = await tx<DeathRecordRow[]>`
          insert into death_records
            (farm_id, animal_id, reported_by_user_id, cause_of_death, date_of_death, image_path, image_hash, notes)
          values
            (${user.farm_id}, ${animalId}, ${user.id}, ${causeOfDeath}, ${dateOfDeath}, ${imageRef}, ${imageHash}, ${
          notes ?? null
        })
          returning *
        `;
        await tx`update animals set status = 'dead', updated_at = now() where id = ${animalId}`;
        return [rec];
      });
    } catch (err) {
      if (isUniqueViolation(err)) {
        // DB unique constraints (one record per animal, per-farm image hash)
        // are authoritative under concurrency. If a racing request beat us,
        // drop the object we just wrote and report 409.
        await deleteFn(imageRef);
        return jsonError(
          c,
          409,
          "A death report for this animal or image already exists",
        );
      }
      throw err;
    }

    await recordAudit({
      c,
      actor: user,
      action: "create",
      entityType: "death_record",
      entityId: record.id,
      details: { animal_id: animalId, cause_of_death: causeOfDeath },
    });

    return c.json(await toResponse(record), 201);
  });

  router.get("/:id", requireAuth, async (c) => {
    const id = Number(c.req.param("id"));
    if (!Number.isInteger(id)) {
      return jsonError(c, 422, "id: Invalid record id");
    }

    const user = c.get("user");
    const sql = getDb();
    const [record] = await sql<DeathRecordRow[]>`
      select * from death_records where id = ${id} and farm_id = ${user.farm_id}
    `;
    if (!record) return jsonError(c, 404, "Death record not found");

    if (user.role !== "manager" && record.reported_by_user_id !== user.id) {
      return jsonError(
        c,
        403,
        "You do not have permission to view this record",
      );
    }

    return c.json(await toResponse(record));
  });

  return router;
}

export const deathsRouter = createDeathsRouter();
